src/compiler/checker.ts

```diff
@@ -4703,6 +4703,12 @@ namespace ts {
             return undefined;
         }
 
+        const enum JSThisAssignmentLocation {

```

Comment:
I'm not sure this really is neater than two booleans. Here's the diff to convert to booleans:
```
-            let definedIn = JSThisAssignmentLocation.None;
+            let definedInConstructor = false, definedInMethod = false;

-                    definedIn |= isDeclarationInConstructor(expression) ? JSThisAssignmentLocation.Constructor : JSThisAssignmentLocation.Method;
+                    if (isDeclarationInConstructor(expression)) { definedInConstructor = true; } else { definedInMethod = true; }

-                let constructorTypes = definedIn & JSThisAssignmentLocation.Constructor ? getConstructuredDefinedThisAssignmentTypes(types!, symbol.declarations) : undefined;
+                let constructorTypes = definedInConstructor ? getConstructuredDefinedThisAssignmentTypes(types!, symbol.declarations) : undefined;

-                if (definedIn & JSThisAssignmentLocation.Method) {
+                if (definedInMethod) {


-                        definedIn |= JSThisAssignmentLocation.Constructor;
+                        definedInConstructor = true;

-            const widened = getWidenedType(addOptionality(type, definedIn === JSThisAssignmentLocation.Method));
+            const widened = getWidenedType(addOptionality(type, definedInMethod && !definedInConstructor));
```

Reply:
Fair enough. I started getting rid of booleans and couldn't stop. :)